fendermint/vm/topdown/src/sync/syncer.rs

```diff
@@ -118,11 +107,54 @@ where
         atomically(|| self.provider.cached_blocks()).await > max_cache_blocks
     }
 
-    /// Poll the next block height. Returns finalized and executed block data.
-    async fn poll_next(&mut self) -> Result<(), Error> {
-        let height = self.sync_pointers.head() + 1;
-        let parent_block_hash = self.non_null_parent_hash().await;
+    /// Get the data needed to pull the next block
+    async fn get_from_cache(&self) -> (BlockHeight, BlockHash) {
+        // we are getting the latest height fetched in cache along with the first non null block
+        // that is stored in cache.
+        // we are doing two fetches in one `atomically` as if we get the data in two `atomically`,
+        // the cache might be updated in between the two calls. `atomically` should guarantee atomicity.
+        atomically(|| {
+            let latest_height = if let Some(h) = self.provider.latest_height()? {
+                h
+            } else {
+                unreachable!("guaranteed to have last committed finality, report bug please")
+            };
+
+            // first try to get the first non null block before latest_height + 1, i.e. from cache
+            let prev_non_null_height =
+                if let Some(height) = self.provider.first_non_null_block(latest_height)? {
+                    tracing::debug!(height, "first non null block in cache");
+                    height
+                } else if let Some(p) = self.provider.last_committed_finality()? {
+                    tracing::debug!(
+                        height = p.height,
+                        "first non null block not in cache, use latest finality"
+                    );
+                    p.height
+                } else {
+                    unreachable!("guaranteed to have last committed finality, report bug please")
+                };
+
+            let hash = if let Some(h) = self.provider.block_hash(prev_non_null_height)? {
+                h
+            } else {
+                unreachable!(
+                    "guaranteed to have hash as the height {} is found",
+                    prev_non_null_height
+                )
+            };
 
+            Ok((latest_height, hash))
+        })
+        .await
+    }
+
+    /// Poll the next block height. Returns finalized and executed block data.
```

Comment:
Maybe it is worth adding a comment here and be explicit about the height and data returned. 
- We get `height` and return the state at `parent_block` right?